sw/device/silicon_creator/lib/shutdown.c

```diff
@@ -0,0 +1,258 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/shutdown.h"
+
+#include <assert.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/stdasm.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/silicon_creator/lib/base/abs_mmio.h"
+#include "sw/device/silicon_creator/lib/drivers/alert.h"
+#include "sw/device/silicon_creator/lib/drivers/lifecycle.h"
+
+#include "alert_handler_regs.h"
+#include "flash_ctrl_regs.h"
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "keymgr_regs.h"
+#include "otp_ctrl_regs.h"
+#include "sram_ctrl_regs.h"
+
+static_assert(ALERT_HANDLER_ALERT_CLASS_MULTIREG_COUNT <=
+                  OTP_CTRL_PARAM_ROM_ALERT_CLASSIFICATION_SIZE / 4,
+              "More alerts than alert classification OTP words!");
+
+#ifdef MOCK_ABS_MMIO
+// If we're building as a unit test, rename the shutdown functions so they
+// can be mocked and/or tested individually.
+// The unmodified function name will be declared as `extern` so the test
+// program can supply its own implementation.  The implementation present
+// in this file will be named `internal_${name}` so the test program can
+// invoke it for testing.
+#define SHUTDOWN_FUNC(name_) \
+  extern void name_;         \
+  void internal_##name_
+#else
+#define SHUTDOWN_FUNC(name_) static ALWAYS_INLINE void name_
+#endif
+
+// TODO: use the real OTP driver after it's converted to abs_mmio.
+#ifdef MOCK_ABS_MMIO
+extern uint32_t otp_read32(uint32_t address);
+#else
+inline uint32_t otp_read32(uint32_t address) {
+  return abs_mmio_read32(TOP_EARLGREY_OTP_CTRL_BASE_ADDR +
+                         OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET + address);
+}
+#endif
+
+// Convert the alert class to an index.
+// This is required because I expect to change the constant definitions in
+// alert_class_t to have reasonable hamming distances.
+static size_t clsindex(alert_class_t cls) {
+  switch (cls) {
+    case kAlertClassA:
+      return 0;
+    case kAlertClassB:
+      return 1;
+    case kAlertClassC:
+      return 2;
+    case kAlertClassD:
+      return 3;
+    default:
+      return 0;
+  }
+}
+
+rom_error_t shutdown_init(lifecycle_state_t lc_state) {
+  rom_error_t error = kErrorOk;
+
+  // Are we in a lifecycle state which needs alert configuration?
+  uint32_t shift;
+  switch (lc_state) {
+    case kLcStateProd:
+      shift = 0;
+      break;
+    case kLcStateProdEnd:
+      shift = 8;
+      break;
+    case kLcStateDev:
+      shift = 16;
+      break;
+    case kLcStateRma:
+      shift = 24;
+      break;
+    default:
+      // No alert init for any other lifecycle states.
+      return error;
```

Comment:
Nit: I'd return kOk here and move the definition of `error` down to the loop where you use it.

Reply:
Done.